include/seahorn/PredicateAbstraction.hh

```diff
@@ -0,0 +1,151 @@
+#ifndef PREDICATE_ABSTRACTION__HH_
+#define PREDICATE_ABSTRACTION__HH_
+
+#include "llvm/Pass.h"
+#include "llvm/IR/Module.h"
+#include "seahorn/HornClauseDB.hh"
+#include "seahorn/HornifyModule.hh"
+
+#include "ufo/Expr.hpp"
+#include "ufo/Smt/Z3n.hpp"
+#include "ufo/Smt/EZ3.hh"
+#include "seahorn/HornClauseDBWto.hh"
+
+namespace seahorn
+{
+	using namespace llvm;
+
+	class HornDbModel
+	{
+	private:
+		ExprMap relToDefMap;
+		std::map<Expr, ExprVector> currentCandidates;
+	public:
+		HornDbModel() {}
+		void initModelFromFP(HornClauseDB &db, ZFixedPoint<EZ3> &fp);
+		void addDef(Expr fapp, Expr def);
+		Expr getDef(Expr fapp);
+		HornDbModel(ExprMap model) : relToDefMap(model) {}
+		virtual ~HornDbModel(){}
+		void setCurrentCandidates(std::map<Expr, ExprVector> currentCandidates) { this->currentCandidates = currentCandidates; }
+//		ExprMap& getRelToSolutionMap() {return relToDefMap;}
+//		void setRelToSolutionMap(ExprMap map) {relToDefMap = map;}
+	};
+
+	class HornModelConverter
+	{
+	public:
+		// converts a model from one database to another. returns false on failure.
+		virtual bool convert (HornDbModel &in, HornDbModel &out) = 0;
+	};
+
+	class PredAbsHornModelConverter : public HornModelConverter
+	{
+	private:
+		std::map<Expr, ExprMap> relToBoolToTermMap;
+		std::map<Expr, Expr> newToOldPredMap;
+		HornClauseDB* abs_db;
+	public:
+		PredAbsHornModelConverter() {}
+		bool convert (HornDbModel &in, HornDbModel &out);
+		std::map<Expr, ExprMap>& getRelToBoolToTermMap() {return relToBoolToTermMap;}
```

Comment:
the only public method should be `convert`. The rest are implementation detail that should not be visible to clients.